### PR TITLE
add support for list, dict, bool, and dataclass in pyflyte run

### PR DESCRIPTION
**Pull request by cosmicBboy**
Signed-off-by: Niels Bantilan <niels.bantilan@gmail.com>

# TL;DR
- Add support for lists and dicts as json string inputs in `pyflyte run`
- Convert `--wf-kwarg` to `wf_kwarg` so that `--wf-kwarg` is a valid way of specifying workflow inputs.
- Try to fetch workflow from remote and register it if it doesn't exist.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

When a workflow uses a `list` or `dict` as input, this PR adds support for specifying a json string input in `pyflyte run`.

Consider the script:

```python
# testing_list_pyflyte_run.py
from typing import List

from flytekit import task, workflow

@task
def process_list(my_list: List[int]) -> int:
    return sum(my_list)

@workflow
def wf(my_list: List[int]) -> int:
    return process_list(my_list=my_list)
```

This this should be a valid `pyflyte run` call:
```
pyflyte run testing_list_pyflyte_run.py:wf --remote --my_list '[1, 2, 3]'
```

When trying to execute the workflow, Flyte handles type checking, so this won't work:
```
pyflyte run testing_list_pyflyte_run.py:wf --remote --my_list '[1, "2", 3]'
```

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: flytekit/clis/sdk_in_container/run.py [hunk 1]
```diff
@@ -189,6 +191,8 @@ def _parse_workflow_inputs(click_ctx, wf_entity, create_upload_location_fn: Opti
             value = int(value)
         elif python_type == float:
             value = float(value)
+        elif python_type.__origin__ in {list, dict}:
+            value = json.loads(value)
```
Comment (cosmicBboy) [inline flytekit/clis/sdk_in_container/run.py:195]: in this case, the type engine makes sure this is the correct type when the workflow is executed.

Should there be some up-front type-checking at this point?

File: flytekit/clis/sdk_in_container/run.py [hunk 2]
```diff
@@ -130,14 +132,23 @@ def run(
...
-        )
+
+        try:
+            wf = remote.fetch_workflow(
+                project=project,
+                domain=domain,
+                name=wf_entity.name,
+                version=version,
+            )
+        except FlyteEntityNotExistException:
```
Comment (kumare3) [inline flytekit/clis/sdk_in_container/run.py:143]: Why do we need this? Why not catch already exists

File: flytekit/clis/sdk_in_container/run.py (hunk 1 above)
Comment (kumare3) [reply flytekit/clis/sdk_in_container/run.py:206]: I guess that is good enough

File: flytekit/core/type_engine.py [hunk 3]
```diff
@@ -481,6 +482,7 @@ def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type:
         dc = cast(DataClassJsonMixin, expected_python_type).from_json(_json_format.MessageToJson(lv.scalar.generic))
         return self._fix_dataclass_int(expected_python_type, self._deserialize_flyte_type(dc, expected_python_type))
 
+    @lru_cache(typed=True)
```
Comment (cosmicBboy) [inline flytekit/core/type_engine.py:485]: added this since the `run` command needs to call `guess_python_type` to get the TypeEngine-derived dataclass. Without caching here, a separate call to `guess_python_type` in the `remote.execute` call would result in a logically equivalent (but new) dataclass, which `TypeEngine.assert_type` would not be happy about.

File: flytekit/clis/sdk_in_container/run.py (hunk 2 above)
Comment (cosmicBboy) [reply flytekit/clis/sdk_in_container/run.py:144]: doing it the other way around would produce a `_InactiveRpcError`
```
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INVALID_ARGUMENT
	details = "task with different structure already exists with id resource_type:TASK project:"flytesnacks" domain:"development" name:"testing_list_pyflyte_run.add_data" version:"e264ed653a92029fe59b53ac62fd9510""
	debug_error_string = "{"created":"@1649691831.919710000","description":"Error received from peer ipv6:[2606:4700:3036::6815:2f8d]:443","file":"src/core/lib/surface/call.cc","file_line":904,"grpc_message":"task with different structure already exists with id resource_type:TASK project:"flytesnacks" domain:"development" name:"testing_list_pyflyte_run.add_data" version:"e264ed653a92029fe59b53ac62fd9510"","grpc_status":3}"
>
```

Unless I'm missing something, this way is cleaner since a `FlyteEntityNotExistException` is actually raised and can be cleanly caught in the `except` clause.

File: flytekit/clis/sdk_in_container/run.py [hunk 4]
```diff
@@ -189,6 +202,15 @@ def _parse_workflow_inputs(click_ctx, wf_entity, create_upload_location_fn: Opti
...
         elif python_type == float:
             value = float(value)
+        elif python_type == bool:
+            value = value in {"True", "true", "1"}
+        elif python_type == datetime:
+            value = datetime.fromtimestamp(int(value)) if value.isnumeric() else datetime.fromisoformat(value)
+        elif getattr(python_type, "__origin__", None) in {list, dict}:
+            value = json.loads(value)
+        elif is_dataclass(python_type):
+            python_type_from_type_engine = TypeEngine.guess_python_type(wf_entity.interface.inputs[argument].type)
```
[the remark this answers is not in the document]
Comment (kumare3) [reply flytekit/clis/sdk_in_container/run.py:212]: I think you can simply do
python_type(**json.loads(value)) ?

[the remark this answers is not in the document]
Comment (eapolinario) [reply flytekit/clis/sdk_in_container/run.py:212]: I was thinking of just passing it to `json.loads` as well.

File: flytekit/clis/sdk_in_container/run.py [hunk 5]
```diff
@@ -189,6 +202,15 @@ def _parse_workflow_inputs(click_ctx, wf_entity, create_upload_location_fn: Opti
             value = int(value)
         elif python_type == float:
             value = float(value)
+        elif python_type == bool:
+            value = value in {"True", "true", "1"}
```
Comment (eapolinario) [inline flytekit/clis/sdk_in_container/run.py:206]: how about the false equivalent values?

File: flytekit/clis/sdk_in_container/run.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cosmicBboy) [reply flytekit/clis/sdk_in_container/run.py:212]: that doesn't work... maybe someone more familiar with the type engine can debug it
```
    return f(get_current_context(), *args, **kwargs)
  File "/Users/nielsbantilan/git/flytekit/flytekit/clis/sdk_in_container/run.py", line 156, in run
    execution = remote.execute(
  File "/Users/nielsbantilan/git/flytekit/flytekit/remote/remote.py", line 744, in execute
    return self.execute_remote_wf(
  File "/Users/nielsbantilan/git/flytekit/flytekit/remote/remote.py", line 837, in execute_remote_wf
    return self.execute_remote_task_lp(
  File "/Users/nielsbantilan/git/flytekit/flytekit/remote/remote.py", line 810, in execute_remote_task_lp
    return self._execute(
  File "/Users/nielsbantilan/git/flytekit/flytekit/remote/remote.py", line 612, in _execute
    literal_inputs = TypeEngine.dict_to_literal_map(ctx, inputs, type_hints)
  File "/Users/nielsbantilan/git/flytekit/flytekit/core/type_engine.py", line 768, in dict_to_literal_map
    raise user_exceptions.FlyteTypeException(type(v), python_type, received_value=v)
flytekit.exceptions.user.FlyteTypeException: Type error!  Received: <class 'testing_list_pyflyte_run.PandasData'> with value: PandasData(id=5, name='foo'), Expected: <class 'types.PandasdataSchema'>.
```

File: flytekit/clis/sdk_in_container/run.py [hunk 6]
```diff
@@ -155,7 +168,8 @@ def run(
             wf_entity,
             is_remote=False,
         )
-        click.secho(wf_entity(**inputs))
+        output = wf_entity(**inputs)
+        click.echo(output)
```
Comment (cosmicBboy) [inline flytekit/clis/sdk_in_container/run.py:172]: click.secho can't handle printing dataframe objects 🤷‍♂️

**Review by eapolinario — APPROVED** [review]

**Review by kumare3 — APPROVED** [review]
